doc/debugger.rst

```diff
@@ -1773,6 +1809,105 @@ target endianness and interpret the bytecode based on that.
+message and the dvalues it contains are entirely up to the implementation and
+depending on the needs of the application, need not be supported at all.
+
+See "Custom requests and notifications" below for more details.
+
+Custom requests and notifications
+=================================
+
+Starting in Duktape 1.5.x, Duktape supports direct communication between the
+debug client and debug target over the same transport by using the special
+'AppRequest' and 'AppNotify' messages. These messages have no meaning to
+Duktape, which merely serves to marshal them back and forth through a defined
+API.
+
+AppNotify messages may be sent by pushing the contents of the message to the
+stack and calling `duk_debugger_notify()` passing the number of values pushed.
+Each value pushed will be sent as a dvalue in the message. So if you push two
+strings, "foo" and "bar", the client will see ``NFY 7 "foo" "bar" EOM``.
+
+AppRequests are used to make requests to the target which are not directly
+related to Ecmascript execution. For example, an AppRequest might be used to:
+
+* Change the frame rate of a game engine
+* Reset/reboot an embedded target device over the debug protocol
+* Perform or trigger software or script updates
+
+Supporting this requires attaching using the ``duk_debugger_attach_custom()``
+API and providing a request callback. When an AppRequest is received, the
+request callback is invoked with the contents of the message on the value
+stack, and may push its own values to be sent in reply.
+
+This is a do-nothing request callback::
+
+    duk_idx_t duk_cb_debug_request(void* udata, duk_context* ctx, duk_idx_t nvalues) {
+	    /* nop */
+	    return 0;
+    }
+
+This simply responds with ``REP EOM`` (an empty reply) to all requests. This is
+not very useful and a real implementation should process the values it receives
+on the stack, push its own values to send in reply, and return a non-negative
+integer indicating how many values it pushed.
+
+Here is a slightly more useful implementation::
+
+    duk_idx_t duk_cb_debug_request(void* udata, duk_context* ctx, duk_idx_t nvalues) {
+	    /* Must use negative stack indices to access dvalues */
+	    const char* cmdname = duk_get_string(ctx, -nvalues + 0);
+	    
+	    if (strcmp(cmdname, "GetVersionInfo") == 0) {
+            /* Return a positive integer to send a REP containing values pushed
+             * to the stack. The return value indicates how many dvalues you
+             * are including in the response.
+             */
+            
+            duk_push_string(ctx, "My Awesome Program");
+            duk_push_number(ctx, 8.12);
+            return 2;  /* 2 dvalues */
+	    } else {
+            /* Return -1 to send an ERR reply. The value on top of the stack
+             * will be string coerced and used for the error text.
+             */
+            
+            duk_push_string(ctx, "unrecognized command name");
+            return -1;
```

Comment:
I'll need to sleep on this error mechanism: negative error returns have an existing meaning in Duktape/C functions and this is a different convention which might be confusing. It still seems better to have an error mechanism that not, so it's probably the best option at this point.

Concretely, it'd be nice if the return values here matched eventual multi-return behavior: the positive >1 return values here would, because they just indicate that multiple return values are on the stack. That negative return is similar to e.g. `DUK_RET_TYPE_ERROR` return value shorthands, but those don't have an associated value stack argument.

Anyway, let me mull this over a bit :)